lib/Model.js

```diff
@@ -276,75 +279,83 @@ function validKeyValue(value) {
   return value !== undefined && value !== null && value !== '';
 }
 
-Model.prototype.put = function(options, next) {
-  debug('put', this);
-  const deferred = Q.defer();
-  let model, model$, item;
+Model.prototype.put = async function(options = {}, next) {
 
-  function putItem() {
-    model$.base.ddb().putItem(item, function(err) {
-      if(err) {
-        deferred.reject(err);
-      }
-      deferred.resolve(model);
-    });
+  if(typeof options === 'function') {
+    next = options;
+    options = {};
   }
 
-  try {
-    options = options || {};
-    if(typeof options === 'function') {
-      next = options;
-      options = {};
-    }
-    if(options.overwrite === null || options.overwrite === undefined) {
-      options.overwrite = true;
-    }
+  if(next){
```

Comment:
Instead of doing this. Why not create a helper function to use throughout the codebase? For example.

```
function promisify(promise, cb) {
    promise.then(result => { cb(null, result); }, err => { next(err); });
    return promise;
}
```

Then instead of this `if` statement. We can just set the promise to a variable instead of returning it. Then do something like:

```
return promisify(promise, next);
```

That way we don't have to write the `then` logic and all of that over and over.

You might have to play with that code. Probably do need a `if` statement within that function I created to ensure that `cb` exists. Not sure if the naming works and everything like that. So might need to play with it quite a bit.

With the current implementation I think someone trying to combine a callback with a promise would fail. So something like:

```
const item = Model.get("myitem", function(err, result) {
    console.log(result);
});
```

I think in that case `item` is undefined with your changes. But I think in the current version `item` would exist.

I would almost consider that a breaking change. I have no idea how people are using Dynamoose, and although I wouldn't recommend that, I can't be positive that people aren't writing code like that.

I **think** my suggestion above would address that.

Reply:
So this is something I definitely struggled with, and would be up for going a different direction. 

If we return a promise and the User doesn't catch it, Node will throw an unhandled promise rejection. Q promises don't dispatch the same node event, so they just fail silently. So we can return the promise, but if they are only using a callback Node will write to std err, which hopefully would never break anything, but would at least likely raise a few eyebrows to whether something changed. I feel like failing silently definitely isn't the way to go though. If you have any ideas for how you want to handle it, let me know!

Definitely up for abstracting it with the above though. I'll try to implement the above. Unfortunately the way some functions dynamically determine if a param is a config or a callback makes it pretty difficult to do the abstraction outside the function. Would definitely make the code cleaner if support was at least dropped for passing in a cb if omitting options, but that would definitely be a breaking change.

Reply:
Thinking on the above, is the plan for the above to continue to support callbacks, or eventually move to entirely promises? Would it make sense to start deprecating certain things like callbacks, or at least passing callbacks in place of options? Might make any eventual breaking changes to the API a lot easier, but I don't know if that's the eventual plan.

Reply:
Using my first example above. Wouldn't we be catching the error there tho? Isn't the only difference the fact that we are returning the promise and abstracting it a bit more?

As of right now there is no plan to deprecate callbacks. It's something I'd be open to discussing. But as of right now there has been no discussion regarding it.

Reply:
This should be abstracted into it's own file/method now. I've called it `linkPromiseToCallback`, and put it in a new `helpers.js` file. Not in love with either though, so if you want me to change the name/location let me know.